Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp

```diff
@@ -109,34 +204,99 @@ namespace AZ::DocumentPropertyEditor
+                            auto& keyType = keyTypeData->Get(nullptr);
+                            DocumentAdapterPtr reflectionAdapter =
+                                AZStd::make_shared<ReflectionAdapter>(m_reservedElementInstance, keyType);
                             Nodes::Adapter::QueryKey.InvokeOnDomNode(impl->m_adapter->GetContents(), &reflectionAdapter, path);
                             return; // key queried; don't store the actual entry until the DPE handles the QueryKey message
                         }
                     }
                 }
 
-                m_container->StoreElement(m_instance, dataAddress);
+                m_container->StoreElement(m_containerInstance, m_reservedElementInstance);
+                m_reservedElementInstance = nullptr;
 
                 auto containerNode = GetContainerNode(impl, path);
-                Nodes::PropertyEditor::AddNotify.InvokeOnDomNode(containerNode);
                 Nodes::PropertyEditor::ChangeNotify.InvokeOnDomNode(containerNode);
-
                 impl->m_adapter->NotifyResetDocument();
             }
 
             void OnRemoveElement(ReflectionAdapterReflectionImpl* impl, const AZ::Dom::Path& path)
             {
-                AZ_Assert(m_elementInstance != nullptr, "Attempted to remove an element without a defined element instance");
-                m_container->RemoveElement(m_instance, m_elementInstance, impl->m_serializeContext);
+                void* instanceToRemove = nullptr;
+                if (m_elementInstance == nullptr)
+                {
+                    instanceToRemove = m_containerInstance;
+                    auto associativeContainer = m_parentContainer->GetAssociativeContainerInterface();
+                    if (associativeContainer)
+                    {
+                        // Use the override instance if specified since it represents the
+                        // element pair instance that the IDataContainer API is expecting
+                        void* keyInstance = m_elementOverrideInstance ? m_elementOverrideInstance : m_containerInstance;
+                        const AZ::SerializeContext::ClassElement* containerClassElement =
+                            m_parentContainer->GetElement(m_parentContainer->GetDefaultElementNameCrc());
+                        instanceToRemove =
+                            associativeContainer->GetElementByKey(m_parentContainerInstance, containerClassElement, keyInstance);
+                    }
+                }
+                else
+                {
+                    instanceToRemove = m_elementInstance;
+                    auto associativeContainer = m_parentContainer->GetAssociativeContainerInterface();
+                    if (associativeContainer)
+                    {
+                        instanceToRemove = associativeContainer->GetElementByKey(m_parentContainerInstance, nullptr, m_elementInstance);
+                    }
```

Comment:
Is `m_elementInstance` always a pointer to the key in this case when using an associative container?

If that is the case we should document the member to indicate when it represents the value_type pair vs when it is just a key_type instance

Reply:
Yes, because we've been storing the "element instance override" in `m_elementInstance`, if the `ContainerElementOverride` attribute is set in the `LegacyReflectionBridge`. I'll add a comment explaining that.

We can't do the same thing for nested containers because they need their serialized container instance ptr in many cases, but the element override for removal from their parent. I left a comment for that part of it, but I'll revisit and make sure it's clear as well.